### PR TITLE
Fix saving branch as scene saves children without owner set

### LINKED ISSUE #82756: Save branch as scene saves children without owner set
### Godot version

4.2.dev [44e399ed5fa895f760b2995e59788bdb49782666]

### System information

Linux

### Issue description

I have node with a `@tool` script. The script adds child nodes needed only for the editor.

savebranch_bug_1

I'm saving the branch of `CustomNode` as a new scene, and the CSGBox saves too, which is unexpected behaviour:

savebranch_bug_2

### Steps to reproduce

- create a new scene
- add node
- attach script:
```
@tool
extends Node3D

var _box = CSGBox3D.new()

func _enter_tree():
    add_child(_box)
    
func _remove_tree():
    remove_child(_box)
```
- save added node as a new scene
- open saved scene

### Minimal reproduction project

[save_branch_bug.zip](https://github.com/godotengine/godot/files/12797975/save_branch_bug.zip)

1. open the project
2. open `save_branch_bug.tscn` if it is not opened already
3. RMB on the `CustomNode` and select "Save branch as a new scene"
4. Open saved scene

You will see CSGBox in a nodes tree. Expected result: no CSGBox node visible in the tree. To make sure: click on the `_CSGBox3D_<number>` node and change it's position. You will notice two boxes. Should be just one.

### PR DESCRIPTION
**Pull request by marcinn**
Fixes #82756 and preserves fix for #42611

### CODE REVIEW

**Comment by KoBeWi** [comment]
With this, the changes from #61966 might not be necessary anymore.

**Comment by marcinn** [comment]
> With this, the changes from #61966 might not be necessary anymore.

I prefer to separate this fix from reverting the other one. I'm not sure if there is a use case where an internal node has an owner set. I'm afraid that reverting #61966 may (re)introduce some issues. Preserving both for a while will be safer for stability.

**Comment by Rindbee** [comment]
Maybe just pass in the corresponding base (child) nodes and find their owners, if the number of their child nodes is the same (Will it be different?).

File: editor/scene_tree_dock.cpp [hunk 1]
```diff
@@ -2663,12 +2663,17 @@ void SceneTreeDock::_new_scene_from(String p_file) {
 	Node *base = selection.front()->get();
 
 	HashMap<const Node *, Node *> duplimap;
+	HashMap<const Node *, Node *> inverse_duplimap;
 	Node *copy = base->duplicate_from_editor(duplimap);
 
+	for (KeyValue item : duplimap) {
```
Comment (KoBeWi) [inline editor/scene_tree_dock.cpp:2669]:
```suggestion
	for (const KeyValue &item : duplimap) {
```
Although not sure if you can use `const` here`.

File: editor/scene_tree_dock.cpp [hunk 2]
```diff
@@ -2663,12 +2663,17 @@ void SceneTreeDock::_new_scene_from(String p_file) {
...
+		inverse_duplimap[item.value] = (Node *)item.key;
```
Comment (KoBeWi) [inline editor/scene_tree_dock.cpp:2670]:
```suggestion
		inverse_duplimap[item.value] = item.key;
```
This should be unnecessary.

File: editor/scene_tree_dock.h [hunk 3]
```diff
@@ -231,7 +231,7 @@ class SceneTreeDock : public VBoxContainer {
 	virtual void shortcut_input(const Ref<InputEvent> &p_event) override;
 
 	void _new_scene_from(String p_file);
-	void _set_node_owner_recursive(Node *p_node, Node *p_owner);
+	void _set_node_owner_recursive(Node *p_node, Node *p_owner, HashMap<const Node *, Node *> &inverse_duplimap);
```
Comment (KoBeWi) [inline editor/scene_tree_dock.h:234]:
```suggestion
	void _set_node_owner_recursive(Node *p_node, Node *p_owner, HashMap<const Node *, Node *> &p_inverse_duplimap);
```

Comment (AThousandShips) [inline editor/scene_tree_dock.h:234]:
```suggestion
	void _set_node_owner_recursive(Node *p_node, Node *p_owner, HashMap<const Node *, Node *> &r_inverse_duplimap);
```
Use `r_` for return values

File: editor/scene_tree_dock.cpp [hunk 4]
```diff
@@ -2702,13 +2707,16 @@ void SceneTreeDock::_new_scene_from(String p_file) {
 	}
 }
 
-void SceneTreeDock::_set_node_owner_recursive(Node *p_node, Node *p_owner) {
-	if (!p_node->get_owner()) {
-		p_node->set_owner(p_owner);
+void SceneTreeDock::_set_node_owner_recursive(Node *p_node, Node *p_owner, HashMap<const Node *, Node *> &inverse_duplimap) {
```
Comment (AThousandShips) [inline editor/scene_tree_dock.cpp:2710]:
```suggestion
void SceneTreeDock::_set_node_owner_recursive(Node *p_node, Node *p_owner, HashMap<const Node *, Node *> &r_inverse_duplimap) {
```

File: editor/scene_tree_dock.cpp [hunk 5]
```diff
@@ -2702,13 +2707,16 @@ void SceneTreeDock::_new_scene_from(String p_file) {
...
+	if (inverse_duplimap.has(p_node)) {
+		Node *original = inverse_duplimap.get(p_node);
```
Comment (KoBeWi) [inline editor/scene_tree_dock.cpp:2712]:
```suggestion
	HashMap<const Node *, Node *>::Iterator E = inverse_duplimap.find(p_node);
	if (E) {
		Node *original = E.value;
```
Avoids double lookup.

File: editor/scene_tree_dock.h (hunk 3 above)
Comment (KoBeWi) [reply editor/scene_tree_dock.h:234]: But it's not return value.
I think you could make it const 🤔

**Comment by marcinn1** [comment]
Thanks for suggestions :+1: , I'll look at them and improve the code as soon as possible.

**Comment by YuriSizov** [comment]
@marcinn1 @marcinn (not sure which account you're actively using?)

Just bumping to remind you about unresolved review comments. This can't make it into 4.2 at this point, but if you could find time to address the notes above, this should help to move this PR along.

**Comment by marcinn1** [comment]
Thanks. I'd look at this at evening.

**Comment by marcinn1** [comment]
I've rebased the branch, added suggested changes (excl `r_` prefix), tested with MRP. Seems to be ok.

File: editor/scene_tree_dock.h [hunk 6]
```diff
@@ -233,7 +233,7 @@ class SceneTreeDock : public VBoxContainer {
 	virtual void shortcut_input(const Ref<InputEvent> &p_event) override;
 
 	void _new_scene_from(String p_file);
-	void _set_node_owner_recursive(Node *p_node, Node *p_owner);
+	void _set_node_owner_recursive(Node *p_node, Node *p_owner, const HashMap<const Node *, Node *> &inverse_duplimap);
```
Comment (AThousandShips) [inline editor/scene_tree_dock.h:236]:
```suggestion
	void _set_node_owner_recursive(Node *p_node, Node *p_owner, const HashMap<const Node *, Node *> &p_inverse_duplimap);
```
Shouldn't be `r_` but should still have a prefix

File: editor/scene_tree_dock.cpp [hunk 7]
```diff
@@ -2820,13 +2825,18 @@ void SceneTreeDock::_new_scene_from(String p_file) {
 	}
 }
 
-void SceneTreeDock::_set_node_owner_recursive(Node *p_node, Node *p_owner) {
-	if (!p_node->get_owner()) {
-		p_node->set_owner(p_owner);
+void SceneTreeDock::_set_node_owner_recursive(Node *p_node, Node *p_owner, const HashMap<const Node *, Node *> &inverse_duplimap) {
```
[the remark this answers is not in the document]
Comment (marcinn1) [reply editor/scene_tree_dock.cpp]: updated

File: editor/scene_tree_dock.cpp (hunk 1 above)
Comment (KoBeWi) [reply editor/scene_tree_dock.cpp:2787]: The KeyValue here does not even use template.
I checked a few options and it doesn't compile. Not worth it.

File: editor/scene_tree_dock.cpp [hunk 8]
```diff
@@ -2820,13 +2825,18 @@ void SceneTreeDock::_new_scene_from(String p_file) {
...
 }
 
-void SceneTreeDock::_set_node_owner_recursive(Node *p_node, Node *p_owner) {
-	if (!p_node->get_owner()) {
-		p_node->set_owner(p_owner);
+void SceneTreeDock::_set_node_owner_recursive(Node *p_node, Node *p_owner, const HashMap<const Node *, Node *> &p_inverse_duplimap) {
+	HashMap<const Node *, Node *>::ConstIterator E = p_inverse_duplimap.find(p_node);
+
+	if (E) {
+		Node *original = E->value;
```
Comment (KoBeWi) [inline editor/scene_tree_dock.cpp:2832]:
```suggestion
		const Node *original = E->value;
```

**Review by KoBeWi — APPROVED** [review]
There are a couple of unresolved minor comments, but otherwise looks ok.

**Comment by marcinn1** [comment]
> There are a couple of unresolved minor comments, but otherwise looks ok.

Should I add these changes before merging?

**Comment by KoBeWi** [comment]
Yep.

**Comment by marcinn1** [comment]
> Yep.

OK, this evening.

**Comment by marcinn1** [comment]
Updated.

> Ok it is necessary, but in this situations we normally use const_cast<>.

Thanks for reminding me why I abandoned c++ ~20yrs ago ;)

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**